module/documents/activity/cast.mjs

```diff
@@ -24,4 +32,69 @@ export default class CastActivity extends ActivityMixin(CastActivityData) {
       sheetClass: CastSheet
     }, { inplace: false })
   );
+
+  /* -------------------------------------------- */
+  /*  Properties                                  */
+  /* -------------------------------------------- */
+
+  /**
+   * Cached copy of the associated spell stored on the actor.
+   * @type {Item5e|void}
+   */
+  get cachedSpell() {
+    return this.actor?.sourcedItems.get(this.spell.uuid, { legacy: false })
+      ?.find(i => i.getFlag("dnd5e", "cachedFor") === this.relativeUUID);
+  }
+
+  /* -------------------------------------------- */
+  /*  Helpers                                     */
+  /* -------------------------------------------- */
+
+  /**
+   * Prepare the data for the cached spell to store on the actor.
+   * @returns {Promise<object|void>}
+   */
+  async getCachedSpellData() {
+    const originalSpell = await fromUuid(this.spell.uuid);
+    if ( !originalSpell ) return;
+    return originalSpell.clone({
+      effects: [
+        ...originalSpell.effects.map(e => e.toObject()),
+        {
+          _id: this.constructor.ENCHANTMENT_ID,
+          type: "enchantment",
+          name: game.i18n.localize("DND5E.CAST.Enchantment.Name"),
+          img: "systems/dnd5e/icons/svg/activity/cast.svg",
+          origin: this.uuid,
+          changes: this.getSpellChanges()
+        }
+      ],
+      flags: {
+        dnd5e: {
+          cachedFor: this.relativeUUID
+        }
+      },
+      _stats: { compendiumSource: this.spell.uuid }
+    }).toObject();
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Create spell changes based on the activity's configuration.
+   * @returns {object}
+   */
+  getSpellChanges() {
+    const changes = [];
+    for ( const type of ["activation", "duration", "range", "target"] ) {
+      if ( !this[type].override ) continue;
+      const data = this.toObject()[type];
+      delete data.override;
+      changes.push({ key: `system.${type}`, mode: CONST.ACTIVE_EFFECT_MODES.OVERRIDE, value: JSON.stringify(data) });
+    }
+    for ( const property of this.spell.properties ) {
+      changes.push({ key: "system.properties", mode: CONST.ACTIVE_EFFECT_MODES.ADD, value: `-${property}` });
```

Comment:
Probably shouldn't remove concentration. There are some exceptions, but the majority of items still require the caster to concentrate on the spell after casting it from the item. We could hard-code removing VSM only (or non-'`isTag`' components), but Spell Scrolls, for example, only remove Material components, not others.

Reply:
These aren't hard-coded removals, only what is explicitly set in the `Cast` activity.